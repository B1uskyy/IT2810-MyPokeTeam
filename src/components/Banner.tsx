import '../styles/banner.css';
import { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';

interface BannerProps {
  title: string;
  text: string;
  desktopImage: string;
  mobileImage: string;
}

function Banner({ title, text, desktopImage, mobileImage }: BannerProps) {
  const [bgImage, setBgImage] = useState(desktopImage);
  const [isLoaded, setIsLoaded] = useState(false);
  const location = useLocation();

  useEffect(() => {
    const updateBackground = () => {
      if (window.innerWidth <= 768) {
        setBgImage(mobileImage);
      } else {
        setBgImage(desktopImage);
      }
    };

    updateBackground();
    window.addEventListener('resize', updateBackground);

    setIsLoaded(true);

    return () => {
      window.removeEventListener('resize', updateBackground);
    };
  }, [desktopImage, mobileImage, location.pathname]);

  return (
    <header className="banner" style={{ backgroundImage: `url(${bgImage})` }}>
      <div className={`banner-content ${isLoaded ? 'slide-in' : ''}`}>
        <h1>{title}</h1>
        <p>{text}</p>
      </div>
    </header>
  );
}

export default Banner;
