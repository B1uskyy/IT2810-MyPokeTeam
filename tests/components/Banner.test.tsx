import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import Banner from '../../src/components/Banner';

type TestData = {
  title: string;
  text: string;
  rootImage: string;
  teamImage: string;
  mobileImage: string;
};
const createTestData = () => {
  return {
    title: 'Test',
    text: 'Lorem ipsum',
    rootImage: '../../src/assets/banner.webp',
    teamImage: '../../src/assets/banner-team.jpeg',
    mobileImage: '../../src/assets/banner-team-mb.jpeg',
  };
};

let bannerTestData: TestData;
beforeEach(() => {
  bannerTestData = createTestData();
});

describe('Banner component', () => {
  const setViewportSize = (width: number) => {
    Object.defineProperty(window, 'innerWidth', {
      configurable: true,
      writable: true,
      value: width,
    });
    // src:https://stackoverflow.com/questions/45868042/figuring-out-how-to-mock-the-window-size-changing-for-a-react-component-test
    // and chatgpt: "I get error attempt to assign to const or readonly variable"

    window.dispatchEvent(new Event('resize'));
  };
  it('Renders with correct text and image on root path, and desktop', () => {
    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path={'/'}
            element={
              <Banner
                title={bannerTestData.title}
                text={bannerTestData.text}
                desktopImage={bannerTestData.rootImage}
                mobileImage={bannerTestData.mobileImage}
              />
            }
          />
        </Routes>
      </MemoryRouter>,
    );
    expect(screen.getByText('Test'));
    expect(screen.getByText('Lorem ipsum'));
    expect(screen.getByTestId('background')).toHaveStyle(`background-image: url(${bannerTestData.rootImage})`);
  });
  it('Renders with correct text and image on /team path, and desktop', () => {
    render(
      <MemoryRouter initialEntries={['/team']}>
        <Routes>
          <Route
            path={'/team'}
            element={
              <Banner
                title={bannerTestData.title}
                text={bannerTestData.text}
                desktopImage={bannerTestData.teamImage}
                mobileImage={bannerTestData.mobileImage}
              />
            }
          />
        </Routes>
      </MemoryRouter>,
    );
    expect(screen.getByText('Test'));
    expect(screen.getByText('Lorem ipsum'));
    expect(screen.getByTestId('background')).toHaveStyle(`background-image: url(${bannerTestData.teamImage})`);
  });
  it('Renders with correct text and image on /team path, and mobile', () => {
    setViewportSize(750);

    render(
      <MemoryRouter initialEntries={['/team']}>
        <Routes>
          <Route
            path={'/team'}
            element={
              <Banner
                title={bannerTestData.title}
                text={bannerTestData.text}
                desktopImage={bannerTestData.teamImage}
                mobileImage={bannerTestData.mobileImage}
              />
            }
          />
        </Routes>
      </MemoryRouter>,
    );

    expect(screen.getByText('Test'));
    expect(screen.getByText('Lorem ipsum'));
    expect(screen.getByTestId('background')).toHaveStyle(`background-image: url(${bannerTestData.mobileImage})`);
  });

  it('matches snapshot on root path with desktop view', () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path={'/'}
            element={
              <Banner
                title={bannerTestData.title}
                text={bannerTestData.text}
                desktopImage={bannerTestData.rootImage}
                mobileImage={bannerTestData.mobileImage}
              />
            }
          />
        </Routes>
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot on path /team and desktopimage', () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={['/team']}>
        <Routes>
          <Route
            path={'/team'}
            element={
              <Banner
                title={bannerTestData.title}
                text={bannerTestData.text}
                desktopImage={bannerTestData.teamImage}
                mobileImage={bannerTestData.mobileImage}
              />
            }
          />
        </Routes>
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot on path /team and mobile', () => {
    setViewportSize(750);

    const { asFragment } = render(
      <MemoryRouter initialEntries={['/team']}>
        <Routes>
          <Route
            path={'/team'}
            element={
              <Banner
                title={bannerTestData.title}
                text={bannerTestData.text}
                desktopImage={bannerTestData.teamImage}
                mobileImage={bannerTestData.mobileImage}
              />
            }
          />
        </Routes>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
