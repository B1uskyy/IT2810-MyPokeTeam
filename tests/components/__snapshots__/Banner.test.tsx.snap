// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Banner component > matches snapshot on path /team and desktopimage 1`] = `
<DocumentFragment>
  <header
    class="banner"
    data-testid="background"
    style="background-image: url(../../src/assets/banner-team-mb.jpeg);"
  >
    <div
      class="banner-content slide-in"
    >
      <h1>
        Test
      </h1>
      <p>
        Lorem ipsum
      </p>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Banner component > matches snapshot on path /team and mobile 1`] = `
<DocumentFragment>
  <header
    class="banner"
    data-testid="background"
    style="background-image: url(../../src/assets/banner-team-mb.jpeg);"
  >
    <div
      class="banner-content slide-in"
    >
      <h1>
        Test
      </h1>
      <p>
        Lorem ipsum
      </p>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Banner component > matches snapshot on root path with desktop view 1`] = `
<DocumentFragment>
  <header
    class="banner"
    data-testid="background"
    style="background-image: url(../../src/assets/banner-team-mb.jpeg);"
  >
    <div
      class="banner-content slide-in"
    >
      <h1>
        Test
      </h1>
      <p>
        Lorem ipsum
      </p>
    </div>
  </header>
</DocumentFragment>
`;
